Cuda quantized tensors, support for quantize per channel

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59701
* **#58245**
* #59700

Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

HDCharles committed:
Adding support for cuda for quantized tensors

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding support for cuda for quantized tensors"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding support for cuda for quantized tensors"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cuda/make_per_channel_quantized_tensor.cu:
```diff
@@ -0,0 +1,36 @@
+namespace at {
+namespace native {
+
+Tensor make_per_channel_quantized_tensor_cuda(
```
what does this function do?  Is this supposed to be different than quantizing a tensor per channel?  Does this create an empty tensor?

HDCharles replied:
its just a cuda analog for _make_per_channel_quantized_tensor_cpu in native_functions.yaml which is used in a number of tests. I'm basing it off of make_per_tensor_quantized_tensor_cuda.

vkuzo commented on aten/src/ATen/native/quantized/cuda/make_per_channel_quantized_tensor.cu:
```diff
@@ -0,0 +1,36 @@
+          .add_input(self)
+          .build();
+        gpu_kernel(iter, [] GPU_LAMBDA(underlying_t value) -> scalar_t {
+          return scalar_t(value);
```
is this expected?

HDCharles replied:
That's how its done in the per_tensor one I was going off of

supriyar replied:
It seems to be very similar to the per_tensor one. Any way we can de-duplicate the code? Not sure if we need a separate file for per-tensor vs per-channel for just this one function.

HDCharles replied:
I can put them in the same file and deduplicate the code

vkuzo commented on aten/src/ATen/native/quantized/cuda/affine_quantizer.cu:
```diff
@@ -56,6 +55,138 @@ void dequantize_tensor_per_tensor_affine_cuda(
       });
 }
 
+void quantize_tensor_per_channel_affine_cuda(
```
LG!  For the slight differences between integer and floating point zero_point, did we figure out the reason for the differences?  If yes, would be great to document them in the code.

HDCharles replied:
I believe the only difference is whether you add before/after rounding which affects things due to the rule of rounding towards even numbers when the value is .5

So in practice its not a significant difference.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4971,7 +4971,7 @@
   variants: function
   dispatch:
-    CPU: quantize_per_channel_cpu
+    CPU, CUDA: quantize_per_channel_cpu
```
looks wrong to me

HDCharles replied:
I fixed this in a later diff in the stack. This pr was getting long and to make the necessary change, I have to change all calls to quantize_per_channel_cpu

HDCharles committed:
Update on "Adding support for cuda for quantized tensors"


Summary: I still haven't completed testing, but hopefully this PR doesn't have issues with multiple commits getting split up.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented:
this diff adds azure pipelines to CI, cc @seemethere @malfet

malfet commented:
@ezyang , no it does not, but looks like someone enabled AzurePipeines as a signal for this PR
@bigfootjon do you know what might have caused it?

bigfootjon commented:
@malfet, I don't see any runs listed on the page: https://dev.azure.com/pytorch/PyTorch/_build

This is really weird. I'm not sure what triggered it

HDCharles committed:
Update on "Adding support for cuda for quantized tensors"


Summary: I still haven't completed testing, but hopefully this PR doesn't have issues with multiple commits getting split up.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Adding support for cuda for quantized tensors"


Summary: I still haven't completed testing, but hopefully this PR doesn't have issues with multiple commits getting split up.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

ezyang commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -904,16 +885,16 @@ def test_qtensor_reshape(self):
         for device in get_supported_device_types():
-            x = torch.randn((1, 3, 4))
-            qx = torch.quantize_per_tensor(x.to(device), scale=1.0, zero_point=0, dtype=torch.quint8)
+            x = torch.randn((1, 3, 4)).to(device)
```
This is not idiomatic; better is `torch.randn((1, 3, 4), device=device)` which will skip initializing the tensor on CPU and then copying it to CUDA

HDCharles replied:
alright, will fix

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

supriyar commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -267,23 +267,30 @@ def test_qtensor_legacy_new_failure(self):
-        scales = torch.rand(numel)
-        zero_points_int = torch.randint(0, 10, size=(numel,))
-        zero_points_float = torch.randn(numel)
+        scales = torch.rand(numel).to(device)
+        zero_points_int = torch.randint(0, 10, size=(numel,)).to(device)
+        zero_points_float = torch.randn(numel).to(device)
```
Can we directly create these tensors on the device `torch.randn(.., device=device)`

HDCharles replied:
yeah, let me fix that. I think its fixed in the final diff, but it must not have been applied at this point in the stack.

HDCharles replied:
fixed

supriyar commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -267,23 +267,30 @@ def test_qtensor_legacy_new_failure(self):
+    def test_per_channel_qtensor_creation_cpu(self):
+        self._test_per_channel_qtensor_creation(torch.device('cpu'))
+
+    @unittest.skipIf(not TEST_CUDA, "No gpu is not available.")
```
nit: Should be No gpu is available?

HDCharles replied:
another fix which is only applied at the next PR in the stack, let me fix that.

HDCharles replied:
fixed

supriyar commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -151,13 +151,14 @@ Tensor& quantize_tensor_per_channel_affine(
     checkQuantizedTensor<scalar_t>(fn_name, qtensor);
-    checkZeroPoints<underlying_t>(fn_name, zero_points);
+    if(qtensor.device().type() != c10::DeviceType::CUDA){
+      checkZeroPoints<underlying_t>(fn_name, zero_points);
```
Does this work now for cuda tensors?

HDCharles replied:
this bypasses those checks for cuda tensors because the checkZeroPoints code doesn't work for cuda, those checks for cuda tensors are added back in here: https://github.com/pytorch/pytorch/pull/59701

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

raghuramank100 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -116,6 +116,7 @@ Tensor empty_quantized(
                               .merge_memory_format(memory_format);
 
   Tensor output;
+
```
Can this change be reverted so that this file is not modified?

HDCharles replied:
ya, will do

raghuramank100 commented on aten/src/ATen/native/quantized/cuda/make_per_channel_quantized_tensor.cu:
```diff
@@ -0,0 +1,33 @@
+  const Tensor& scales,
+  const Tensor& zero_points,
+  int64_t axis) {
+    Tensor dst = at::_empty_per_channel_affine_quantized(
```
Can we have a single _empty function that can internally call either per_channel or per_tensor affine? If we can do that, we can remove this function.

HDCharles replied:
we do have empty_quantized which can handle both. But the per_(tensor/channel) make functions have different input types which is the main issue. per_channel takes tensors, per_tensor takes double and int.

Also, we already have:
make_per_tensor_quantized_tensor_cpu
make_per_tensor_quantized_tensor_cuda
make_per_channel_quantized_tensor_cpu

so isn't adding make_per_channel_quantized_tensor_cuda the correct move to preserve the structure?

raghuramank100 reviewed (approved):
Looks almost ready to go, please take a look at the comments and address them prior to landing.

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

malfet commented on aten/src/ATen/native/quantized/cuda/make_quantized_tensor.cu:
```diff
@@ -4,17 +4,10 @@
-      self.options().dtype(toQIntType(self.scalar_type())),
-      scale,
-      zero_point);
+static inline void assign_quantized_tensor_cuda(
```
This should fix
`On Windows, the enclosing parent function for an extended __host__ __device__ lambda cannot have internal or no linkage`
```suggestion
inline void assign_quantized_tensor_cuda(
```

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per channel"


Summary: This adds the support for the per_channel quantization, 

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018271](https://our.internmc.facebook.com/intern/diff/D29018271)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018271).
